alf/networks/network.py

```diff
@@ -0,0 +1,150 @@
+            # call that __init__ instead separately since the parent class is also a
+            # subclass of Network. Here just create the class and return.
+            return abc.ABCMeta.__new__(mcs, classname, baseclasses, attrs)
+
+        arg_spec = inspect.getfullargspec(init)
+        if arg_spec.varargs is not None:
+            raise RuntimeError(
+                "%s.__init__ function accepts *args.  This is not allowed." %
+                classname)
+
+        def _capture_init(self, *args, **kwargs):
+            """Captures init args and kwargs and stores them into `_saved_kwargs`."""
+            if len(args) > len(arg_spec.args) + 1:
+                # Error case: more inputs than args.  Call init so that the appropriate
+                # error can be raised to the user.
+                init(self, *args, **kwargs)
+            for i, arg in enumerate(args):
+                # Add +1 to skip `self` in arg_spec.args.
+                kwargs[arg_spec.args[1 + i]] = arg
+            init(self, **kwargs)
+            setattr(self, "_saved_kwargs", kwargs)
+
+        attrs["__init__"] = functools.update_wrapper(_capture_init, init)
+        return abc.ABCMeta.__new__(mcs, classname, baseclasses, attrs)
+
+
+@six.add_metaclass(_NetworkMeta)
+class Network(nn.Module):
+    """Base extension to nn.Module to simplify copy operations."""
+
+    def __init__(self, input_tensor_spec, state_spec, name):
+        """Creates an instance of `Network`.
+
+        Args:
+            input_tensor_spec (nested TensorSpec):  Representing the
+                input observations.
+            state_spec (nested TensorSpec): Representing the state needed by
+                the network. Use () if none.
+            name (str): A string representing the name of the network.
+        """
+        super(Network, self).__init__()
+        self._name = name
+        self._input_tensor_spec = input_tensor_spec
+        self._state_spec = state_spec
+
+    @property
+    def name(self):
+        return self._name
+
+    @property
+    def state_spec(self):
+        return self._state_spec
+
+    @property
+    def input_tensor_spec(self):
+        """Returns the spec of the input to the network of type InputSpec."""
+        return self._input_tensor_spec
+
+    def copy(self, **kwargs):
+        """Create a shallow copy of this network.
+
+        **NOTE** Network layer weights are *never* copied.  This method recreates
+        the `Network` instance with the same arguments it was initialized with
+        (excepting any new kwargs).
+
+        Args:
+        **kwargs: Args to override when recreating this network.  Commonly
+            overridden args include 'name'.
+
+        Returns:
+        A shallow copy of this network.
+        """
+        return type(self)(**dict(self._saved_kwargs, **kwargs))
+
+    def __call__(self, inputs, *args, **kwargs):
+        alf.nest.assert_same_structure(inputs, self.input_tensor_spec)
```

Comment:
`inputs` has additional batch dims. We can remove this line for now. My local version has this check.